src/socket.rs

```diff
@@ -0,0 +1,610 @@
+        timestamping,
+        InterfaceTimestampMode::HardwarePTPAll | InterfaceTimestampMode::HardwarePTPRecv
+    ) {
+        socket.driver_enable_hardware_timestamping(interface)?;
+    }
+    socket.set_nonblocking(true)?;
+
+    #[cfg(target_os = "linux")]
+    let errqueue_waiter = crate::raw_socket::err_queue_waiter::ErrQueueWaiter::new(&socket)?;
+
+    Ok(Socket {
+        timestamp_mode: timestamping,
+        socket: AsyncFd::new(socket)?,
+        #[cfg(target_os = "linux")]
+        errqueue_waiter,
+        send_counter: 0,
+        _addr: PhantomData,
+        _state: PhantomData,
+    })
+}
+
+pub fn open_interface_udp6(
+    interface: InterfaceName,
+    port: u16,
+    timestamping: InterfaceTimestampMode,
+) -> std::io::Result<Socket<SocketAddrV6, Open>> {
+    // Setup the socket
+    let socket = RawSocket::open(libc::PF_INET6, libc::SOCK_DGRAM, libc::IPPROTO_UDP)?;
+    socket.reuse_addr()?;
+    socket.ipv6_v6only(true)?;
+    socket.bind(SocketAddrV6::new(Ipv6Addr::UNSPECIFIED, port, 0, 0).to_sockaddr(PrivateToken))?;
+    socket.bind_to_device(interface)?;
+    socket.ipv6_multicast_if(interface)?;
+    socket.ipv6_multicast_loop(false)?;
+    configure_timestamping(&socket, timestamping)?;
+    #[cfg(target_os = "linux")]
+    if matches!(
+        timestamping,
+        InterfaceTimestampMode::HardwarePTPAll | InterfaceTimestampMode::HardwarePTPRecv
+    ) {
+        socket.driver_enable_hardware_timestamping(interface)?;
+    }
+    socket.set_nonblocking(true)?;
+
+    #[cfg(target_os = "linux")]
+    let errqueue_waiter = crate::raw_socket::err_queue_waiter::ErrQueueWaiter::new(&socket)?;
+
+    Ok(Socket {
+        timestamp_mode: timestamping,
+        socket: AsyncFd::new(socket)?,
+        #[cfg(target_os = "linux")]
+        errqueue_waiter,
+        send_counter: 0,
+        _addr: PhantomData,
+        _state: PhantomData,
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{net::IpAddr, str::FromStr};
+
+    #[tokio::test]
+    async fn test_open_udp6() {
+        let mut a = open_interface_udp6(
+            InterfaceName::from_str("lo").unwrap(),
+            5123,
+            super::InterfaceTimestampMode::None,
+        )
+        .unwrap();
+        let mut b = connect_address(
+            SocketAddr::new(IpAddr::V6(Ipv6Addr::LOCALHOST), 5123),
+            GeneralTimestampMode::None,
+        )
+        .unwrap();
+        assert!(b.send(&[1, 2, 3]).await.is_ok());
+        let mut buf = [0; 4];
+        let recv_result = a.recv(&mut buf).await.unwrap();
+        assert_eq!(recv_result.bytes_read, 3);
+        assert_eq!(&buf[0..3], &[1, 2, 3]);
+        assert!(a.send_to(&[4, 5, 6], recv_result.remote_addr).await.is_ok());
+        let recv_result = b.recv(&mut buf).await.unwrap();
+        assert_eq!(recv_result.bytes_read, 3);
+        assert_eq!(&buf[0..3], &[4, 5, 6]);
+    }
+
+    #[tokio::test]
+    async fn test_open_udp4() {
+        let mut a = open_interface_udp4(
```

Comment:
this fails because binding to an interface requires root. I don't want to give up this test, so we'll need to think about how to configure whether the bind should occur

Reply:
I don't get failures for this, where does it fail for you?

Reply:
```
test interface::tests::get_address_ipv6_valid ... ok
test socket::tests::test_open_udp4 ... FAILED
test socket::tests::test_open_ip ... ok
test socket::tests::test_open_udp6 ... FAILED
test socket::tests::test_software_timestamping ... ok
test interface::tests::test_interface_from_str ... ok

failures:

---- socket::tests::test_open_udp4 stdout ----
thread 'socket::tests::test_open_udp4' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', src/socket.rs:531:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- socket::tests::test_open_udp6 stdout ----
thread 'socket::tests::test_open_udp6' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', src/socket.rs:507:10


failures:
    socket::tests::test_open_udp4
    socket::tests::test_open_udp6
```

Reply:
Is this on your local machine? On CI and for me locally it seems to work. Not sure where the difference would come from then honestly.

Reply:
Figured out this was a difference between pre-5.9 and 5.9 onwards kernels. See also https://lore.kernel.org/netdev/m37drhs1jn.fsf@bernat.ch/t/